packages/backend/src/domain/sort-programs.ts

```diff
@@ -0,0 +1,51 @@
+import type { ProgramData } from '@tee/web/src/types'
+import { ProgramAidType, TrackHelpValue } from '@tee/web/src/types'
+
+export const sortPrograms = (
+  programs: ProgramData[],
+  questionnaireRoute: TrackHelpValue
+): ProgramData[] => {
+  return programs.sort((p1, p2) => comparePrograms(p1, p2, questionnaireRoute))
+}
+
+const comparePrograms = (
+  program1: ProgramData,
+  program2: ProgramData,
+  route: TrackHelpValue
+): number => {
+  return Math.sign(getPriority(program1, route) - getPriority(program2, route))
+}
+
+const getPriority = (prog: ProgramData, route: TrackHelpValue): number => {
+  switch (route) {
+    case TrackHelpValue.Unknown:
+      if (isCoachingOrTraining(prog) && isFree(prog)) return 1
+      if (isCoachingOrTraining(prog) && isMaybeFree(prog)) return 2
+      if (isCoachingOrTraining(prog)) return 3
+      if (hasType(ProgramAidType.fund, prog)) return 4
+      if (hasType(ProgramAidType.loan, prog)) return 5
+      if (hasType(ProgramAidType.tax, prog)) return 6
+      return 10
+
+    case TrackHelpValue.Precise:
+      if (hasType(ProgramAidType.fund, prog)) return 1
+      if (hasType(ProgramAidType.loan, prog)) return 2
+      if (hasType(ProgramAidType.tax, prog)) return 3
+      if (isCoachingOrTraining(prog) && !isMaybeFree(prog) && !isFree(prog)) return 4
+      if (isCoachingOrTraining(prog) && isMaybeFree(prog) && !isFree(prog)) return 5
+      if (isCoachingOrTraining(prog) && isFree(prog)) return 6
+      return 10
+  }
+}
+
+const hasType = (aidType: ProgramAidType, program: ProgramData) =>
+  program["nature de l'aide"] == aidType
```

Comment:
suggestion : typer la clé (mais je me rends compte que ce typage n'existe pas encore dans `Publicodes.ts`) ? 

```suggestion
  import { AidKeys } from '@/types' // NDR : A créer du coup
  
  program[AidKeys.AidNature] == aidType
```

... et du coup idem pour les autres entrées

Reply:
Les clés sont typées ([ici](https://github.com/betagouv/transition-ecologique-entreprises-widget/blob/preprod/packages/web/src/types/programTypes.ts#L26)) ! Ce sont les espaces dans les noms qui m'obligent (je crois?) à passer par cette syntaxe.

Reply:
je pense qu'on s'est mal compris, je voulais dire que les valeurs des clés sont effectivement typées mais pas les noms des clés eux-mêmes. Ma proposition visait essentiellement à ne plus avoir de mention "en dur" de `"nature de l'aide"` ailleurs que dans un fichier de typage

Reply:
Pour entrer un peu + dans le détail de ce que j'avais en tête =>

```ts
// programTypes.ts

export enum AidKeys {
  AidId: 'id',
  AidTitle: 'titre',
  AidPromise: 'promesse',
  AidDescription: 'description',
  AidDescriptionLong: 'description longue',
  AidIllustration: 'illustration',
  AidGoals: 'objectifs',
  AidOperator: 'opérateur de contact',
  AidNature: 'nature de l'aide',
  // etc...
}

export interface ProgramData {
  [AidKeys.AidId]: string
  [AidKeys.AidTitle]: string
  [AidKeys.AidPromise]: string
  [AidKeys.AidDescription]: string
  [AidKeys.AidDescriptionLong]?: string
  [AidKeys.AidIllustration]: string
  [AidKeys.AidGoals]: string[]
  [AidKeys.AidOperator]: string
  [AidKeys.AidNature]?: string[]
  [AidKeys.AidNature]: ProgramAidType
  
  // etc...

  "coût de l'accompagnement"?: string
  "durée de l'accompagnement"?: string
  'montant du financement'?: string
  "montant de l'avantage fiscal"?: string
  'montant du prêt'?: string
  'durée du prêt'?: string
  'taux du prêt'?: string
  conditions?: Condition[]
  url?: string

  publicodes: Object
}

```

Reply:
J'ai mis un [commentaire](https://github.com/betagouv/transition-ecologique-entreprises-widget/issues/368#issuecomment-1843186663) dans l'issue #368. Est-ce que ça t'irait que je repousse ces modifications à la résolution de cette issue ?

Reply:
partons là-dessus, on peut considérer cette discussion comme `resolved` en attente de #368